staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 // Run spawns the secure http server. It only returns if stopCh is closed
 // or the secure port cannot be listened on initially.
 func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
-	delayedStopCh := make(chan struct{})
+	events := s.TerminationEvents
+
+	signalFn := s.overrideTerminationEventSignalFunc
+	if signalFn == nil {
+		signalFn = func() {
+			// wait until we receive a KILL signal
+			<-stopCh
+			events.WithName(ShutdownInitiated).Signal()
+
+			// fire AfterShutdownDelayDuration event when ShutdownDelayDuration elapses
+			<-time.After(s.ShutdownDelayDuration)
+			events.WithName(AfterShutdownDelayDuration).Signal()
+
+			// wait until all existing request(s) in flight that we account for are drained.
+			s.HandlerChainWaitGroup.Wait()
+			events.WithName(InFlightRequestsDrained).Signal()
+		}
+	}
+	go signalFn()
 
 	go func() {
-		defer close(delayedStopCh)
-
-		<-stopCh
-
 		// As soon as shutdown is initiated, /readyz should start returning failure.
-		// This gives the load balancer a window defined by ShutdownDelayDuration to detect that /readyz is red
-		// and stop sending traffic to this server.
-		close(s.readinessStopCh)
+		// This gives the load balancer a window defined by ShutdownDelayDuration to
+		// detect that /readyz is red and stop sending traffic to this server.
+		defer close(s.readinessStopCh)
 
-		time.Sleep(s.ShutdownDelayDuration)
+		<-events.WithName(ShutdownInitiated).Signaled()
 	}()
 
-	// close socket after delayed stopCh
-	stoppedCh, err := s.NonBlockingRun(delayedStopCh)
+	// Start the audit backend before any request comes in. This means we must call Backend.Run
+	// before http server start serving. Otherwise the Backend.ProcessEvents call might block.
+	if s.AuditBackend != nil {
+		if err := s.AuditBackend.Run(events.WithName(InFlightRequestsDrained).Signaled()); err != nil {
```

Comment:
I think that `InFlightRequestsDrained` doesn't guarantee that all in-flight requests have been processed

Reply:
This is in keeping with the current behavior - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L419-L420

I think this is correct - `WithWaitGroup` precedes `WithAudit` - so existing requests in flight have already been added to the audit backend. so `auditBackend.Run` will stop as soon `WithWaitGroup` has finished executing all existing requests in flight, thoughts?